nmi_mysql/nmi_mysql.py

```diff
@@ -1,194 +1,122 @@
+            conf['password'] + '@',
+            conf['host'] + ':',
+            str(conf['port']) + '/',
+            conf['db'],
+            '?charset=' + self.charset
+        ])
+
+    def _generate_query(self, _query, _params):
         query = re.sub('\?', '%s', _query)
 
         if not isinstance(_params, list):
-            return query % self.to_string(_params)
+            return (query % self._to_string(_params), _params)
 
         params = []
-        values = []
+        multi_insert = []
+        other_queries = []
+
         for param in _params:
+            if isinstance(param, list):
+                params.extend(param)
+
+            elif isinstance(param, dict):
+                for key in param:
+                    params.append(param[key])
+
+            else:
+                params.append(param)
+
             if isinstance(param, tuple):
-                values.append('(' + self.to_string(param) + ')')
+                multi_insert.append(self._to_string(param))
 
             else:
-                params.append(self.to_string(param))
+                other_queries.append(self._to_string(param))
 
-        if values:
-            params = ', '.join(values)
-            query = query % params[1:-1]
+        if multi_insert:
+            return (query % ', '.join(multi_insert), params)
 
         else:
-            query = query % tuple(params)
-
-        return query
+            return (query % tuple(other_queries), params)
 
-    def to_string(self, temp):
-        if isinstance(temp, (list, tuple)):
+    def _to_string(self, temp):
+        if isinstance(temp, list):
             tmp = []
             for item in temp:
-                if isinstance(item, str):
-                    item = item.replace('%', '%%')
-                tmp.append(self.handle.escape(item))
+                tmp.append(self._to_string(item))
 
             return ', '.join(tmp)
 
         elif isinstance(temp, dict):
             tmp = []
             for key in temp:
-                if isinstance(temp[key], str):
-                    temp[key] = temp[key].replace('%', '%%')
-                tmp.append(key + ' = ' + self.handle.escape(temp[key]))
+                tmp.append(key + ' = ' + self._to_string(temp[key]))
 
             return ', '.join(tmp)
 
-        elif isinstance(temp, str):
-            return self.handle.escape(temp.replace('%', '%%'))
-
-        else:
-            return self.handle.escape(temp)
-
+        return '%s'
 
-class ConnectionPool(object):
-    """
-    Usage:
-        conn_pool = nmi_mysql.ConnectionPool(config)
-
-        db = conn_pool.get_connection()
-        db.query('SELECT 1', [])
-        conn_pool.return_connection(db)
-
-        conn_pool.close()
-    """
-
-    def __init__(self, conf, max_pool_size=20):
-        self.conf = conf
-        self.max_pool_size = max_pool_size
-        self.initialize_pool()
+    def connect(self):
+        self.con = self.engine.connect()
```

Comment:
Supporting retry is a good feature. However, I would disagree on default retry, it masks too many necessary error during development.

Please make sure retry is zero (ie no retry) unless the user specifies it.